packages/salesforce-adapter/src/filters/field_permissions_enum.ts

```diff
@@ -0,0 +1,282 @@
+          return fieldPermissionsObjectToEnum(fieldPermission)
+        }
+        return fieldPermission
+      }
+    )
+  )
+  return instance
+}
+
+const fieldPermissionValuesToObject = (instance: InstanceElement): InstanceElement => {
+  const { fieldPermissions } = instance.value
+  if (fieldPermissions === undefined || !_.isPlainObject(fieldPermissions)) {
+    log.warn('Instance of type %s does not have fieldPermissions value or is not an object (as expected)', instance.elemID.typeName)
+    return instance
+  }
+  instance.value.fieldPermissions = _.mapValues(
+    fieldPermissions,
+    (objectPermission, objectName) => _.mapValues(
+      objectPermission,
+      (fieldPermissionValue, fieldName) => (
+        permissionsEnumToObject(`${objectName}.${fieldName}`, fieldPermissionValue)
+      )
+    )
+  )
+  return instance
+}
+
+const fieldPermissionFieldToEnum = (objectType: ObjectType): void => {
+  if (isTypeWithFieldPermissions(objectType)) {
+    objectType.fields.fieldPermissions.refType = getMapOfMapOfFieldPermissionEnum()
+  }
+}
+
+const fieldPermissionsFieldToOriginalType = (objectType: ObjectType): void => {
+  if (isTypeWithFieldPermissions(objectType)) {
+    objectType.fields.fieldPermissions.refType = getMapOfMapOfProfileFieldLevelSecurity()
+  }
+}
+
+let shouldRunDeployFilters: boolean
+
+// The decision if to run the deploy filters is based on the fieldPermissions type
+// which indicates if it this filter ran onFetch or not
+const shouldRunDeployFiltersAccordingToInstanceType = async (
+  instanceType: ObjectType,
+): Promise<boolean> =>
+  (isTypeWithFieldPermissions(instanceType)
+    && (await instanceType.fields.fieldPermissions.getType()).elemID
+      .isEqual(getMapOfMapOfFieldPermissionEnum().elemID))
+
+
+const filter: LocalFilterCreator = ({ config }) => ({
+  onFetch: async elements => {
+    if (config.enumFieldPermissions === false) {
+      return
+    }
+    log.info('Running enumFieldPermissions onFetch')
+    await awu(metadataTypesWithFieldPermissions).forEach(async metadataType => {
+      const type = await awu(elements).find(
+        async element => isObjectType(element) && await apiName(element) === metadataType,
+      ) as ObjectType
+      if (type !== undefined) {
+        fieldPermissionFieldToEnum(type)
+      }
+    })
+    const instancesWithFieldPermissions = await awu(elements)
+      .filter(isInstanceOfType(...metadataTypesWithFieldPermissions))
+      .toArray() as InstanceElement[]
+    instancesWithFieldPermissions.forEach(fieldPermissionValuesToEnum)
+    elements.push(enumFieldPermissions)
+  },
+  preDeploy: async changes => {
+    const instanceChangesWithFieldPermissions = await awu(changes)
+      .filter(isAdditionOrModificationChange)
+      .filter(isInstanceOfTypeChange(...metadataTypesWithFieldPermissions))
+      .toArray() as (AdditionChange<InstanceElement> | ModificationChange<InstanceElement>)[]
```

Comment:
I wonder if in order to avoid the cast at the end it would be better to do something like
```suggestion
    const instanceChangesWithFieldPermissions = (
      await awu(changes)
        .filter(isInstanceOfTypeChange(...metadataTypesWithFieldPermissions))
        .toArray()
    )
      .filter(isAdditionOrModificationChange)
      .filter(isInstanceChange)
```

Reply:
It's less efficient and the isInstanceChange basically does nothing, is it really worth it?